### PR TITLE
MDS207 | Implement authentication model in conjunction with scope handler pattern

### PR DESCRIPTION
**Pull request by ajchilts**

### CODE REVIEW

**Review by grab-a-byte — APPROVED** [review]
LGTM

**Review by grab-a-byte — COMMENTED** [review]
No Api config class?
Ignore pulled locally and seen its there.

File: app/uk/gov/hmrc/individualsbenefitsandcreditsapi/domains/ErrorResponses.scala [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                          "NOT_FOUND",
+                          "The resource can not be found")
+case object ErrorTooManyRequests
+    extends ErrorResponse(TOO_MANY_REQUESTS,
+                          "TOO_MANY_REQUESTS",
+                          "Rate limit exceeded")
+
+class MatchNotFoundException extends RuntimeException
+class ValidationException(message: String) extends RuntimeException(message)
+class ScopeAuthorisationException(message: String)
```
Comment (owen-gilfellon) [inline app/uk/gov/hmrc/individualsbenefitsandcreditsapi/domains/ErrorResponses.scala]: ScopeAuthorisationException and ValidationException not used

File: app/uk/gov/hmrc/individualsbenefitsandcreditsapi/services/ScopesService.scala [hunk 2]
```diff
@@ -73,4 +73,15 @@ class ScopesService @Inject()(configuration: Configuration) {
...
       .toMap
 
+  def getEndPointScopes(endpointKey: String): Iterable[String] = {
+    val keys = apiConfig
+      .getEndpoint(endpointKey)
+      .map(endpoint => endpoint.fields.keys.toList.sorted)
+      .getOrElse(List())
+
+    apiConfig.scopes
+      .filter(s => s.fields.toSet.subsetOf(keys.toSet))
```
Comment (owen-gilfellon) [inline app/uk/gov/hmrc/individualsbenefitsandcreditsapi/services/ScopesService.scala]: This doesn't work if a scope has fields from multiple endpoints (e.g. a default scope) - definition of this.subsetOf(that) is "true if this set is a subset of that, i.e. if every element of this set is also an element of that".... suggest something like:
`.filter(s => s.fields.toSet.intersect(keys.toSet).nonEmpty)`

**Review by owen-gilfellon — CHANGES_REQUESTED** [review]
Just a couple of small changes. Test coverage is down slightly (mostly the exceptions in CommonController that aren't being triggered), think that can be fixed later though.

**Review by owen-gilfellon — APPROVED** [review]

**Pull request merged**